src/circuit/gadget/ecc/chip/add.rs

```diff
@@ -0,0 +1,478 @@
+    // α = inv0(x_q - x_p)
+    alpha: Column<Advice>,
+    // β = inv0(x_p)
+    beta: Column<Advice>,
+    // γ = inv0(x_q)
+    gamma: Column<Advice>,
+    // δ = inv0(y_p + y_q) if x_q = x_p, 0 otherwise
+    delta: Column<Advice>,
+    // Permutation
+    perm: Permutation,
+}
+
+impl From<&EccConfig> for Config {
+    fn from(ecc_config: &EccConfig) -> Self {
+        Self {
+            q_add: ecc_config.q_add,
+            x_p: ecc_config.advices[0],
+            y_p: ecc_config.advices[1],
+            x_qr: ecc_config.advices[2],
+            y_qr: ecc_config.advices[3],
+            lambda: ecc_config.advices[4],
+            alpha: ecc_config.advices[5],
+            beta: ecc_config.advices[6],
+            gamma: ecc_config.advices[7],
+            delta: ecc_config.advices[8],
+            perm: ecc_config.perm.clone(),
+        }
+    }
+}
+
+impl Config {
+    pub(crate) fn advice_columns(&self) -> HashSet<Column<Advice>> {
+        core::array::IntoIter::new([
+            self.x_p,
+            self.y_p,
+            self.x_qr,
+            self.y_qr,
+            self.lambda,
+            self.alpha,
+            self.beta,
+            self.gamma,
+            self.delta,
+        ])
+        .collect()
+    }
+
+    pub(crate) fn create_gate<F: FieldExt>(&self, meta: &mut ConstraintSystem<F>) {
+        let q_add = meta.query_selector(self.q_add, Rotation::cur());
+        let x_p = meta.query_advice(self.x_p, Rotation::cur());
+        let y_p = meta.query_advice(self.y_p, Rotation::cur());
+        let x_q = meta.query_advice(self.x_qr, Rotation::cur());
+        let y_q = meta.query_advice(self.y_qr, Rotation::cur());
+        let x_r = meta.query_advice(self.x_qr, Rotation::next());
+        let y_r = meta.query_advice(self.y_qr, Rotation::next());
+        let lambda = meta.query_advice(self.lambda, Rotation::cur());
+
+        // α = inv0(x_q - x_p)
+        let alpha = meta.query_advice(self.alpha, Rotation::cur());
+        // β = inv0(x_p)
+        let beta = meta.query_advice(self.beta, Rotation::cur());
+        // γ = inv0(x_q)
+        let gamma = meta.query_advice(self.gamma, Rotation::cur());
+        // δ = inv0(y_p + y_q) if x_q = x_p, 0 otherwise
+        let delta = meta.query_advice(self.delta, Rotation::cur());
+
+        // Useful composite expressions
+        // α ⋅(x_q - x_p)
+        let if_alpha = (x_q.clone() - x_p.clone()) * alpha;
+        // β ⋅ x_p
+        let if_beta = x_p.clone() * beta;
+        // γ ⋅ x_q
+        let if_gamma = x_q.clone() * gamma;
+        // δ ⋅(y_p + y_q)
+        let if_delta = (y_q.clone() + y_p.clone()) * delta;
+
+        // Useful constants
+        let one = Expression::Constant(F::one());
+        let two = Expression::Constant(F::from_u64(2));
+        let three = Expression::Constant(F::from_u64(3));
+
+        // Handle cases in incomplete addition
+        {
+            meta.create_gate(
+                "(x_q − x_p)⋅((x_q − x_p)⋅λ − (y_q−y_p))=0",
```

Comment:
```suggestion
                "(x_q − x_p)⋅((x_q − x_p)⋅λ − (y_q−y_p)) = 0",
```

Reply:
Resolving this (and similar suggested changes below) because we are no longer creating a single gate for each polynomial constraint, as of https://github.com/zcash/halo2/pull/273.